pkg/middleware/gziper.go

```diff
@@ -18,26 +21,59 @@ var gzipIgnoredPathPrefixes = []string{
 	"/api/live/push", // WebSocket does not support gzip compression.
 }
 
-func Gziper() macaron.Handler {
-	gziperLogger := log.New("gziper")
-	gziper := gzip.Gziper()
+type gzipResponseWriter struct {
+	w *gzip.Writer
+	macaron.ResponseWriter
+}
+
+func (grw *gzipResponseWriter) WriteHeader(c int) {
+	grw.Header().Del("Content-Length")
+	grw.ResponseWriter.WriteHeader(c)
+}
+
+func (grw gzipResponseWriter) Write(p []byte) (int, error) {
+	if grw.Header().Get("Content-Type") == "" {
+		grw.Header().Set("Content-Type", http.DetectContentType(p))
+	}
+	grw.Header().Del("Content-Length")
+	return grw.w.Write(p)
+}
+
+func (grw gzipResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	if hijacker, ok := grw.ResponseWriter.(http.Hijacker); ok {
+		return hijacker.Hijack()
+	}
+	return nil, nil, fmt.Errorf("GZIP ResponseWriter doesn't implement the Hijacker interface")
+}
 
-	return func(ctx *macaron.Context) {
-		requestPath := ctx.Req.URL.RequestURI()
+func Gziper() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			requestPath := req.URL.RequestURI()
+
+			for _, pathPrefix := range gzipIgnoredPathPrefixes {
+				if strings.HasPrefix(requestPath, pathPrefix) {
+					return
+				}
+			}
+
+			// ignore resources
+			if (strings.HasPrefix(requestPath, "/api/datasources/") || strings.HasPrefix(requestPath, "/api/plugins/")) && strings.Contains(requestPath, resourcesPath) {
```

Comment:
It seems this is pre-existing, but is `strings.Contains()` really what we want here for the path `/resources`, rather than `strings.HasPrefix()`?

Reply:
This is actually a great catch! I'm not sure if anyone uses Grafana served in a subpath (I've seen related old issues in the past), but you're right - in the existing shape it won't work for subpaths, `Contains()` would work better. 
Until someone gives a UID to their dashboard named "resources" and tries to invoke some API call with that UID.

Reply:
The original behaviour turned out to be correct, and have been introduced in 4ff613a432b6ee75be6a21495c8f414c12aded2b commit. Also there are no issues with Grafana being served from a subpath.
However, to avoid duplicate lookup code I have updated this PR. It now uses a list of path matchers rather than hard-coded logic for individual subpaths. Still ~10x-100x faster than regexps and flexible enough to filter out everything needed.